### PR TITLE
Fix SSO to console through myaccount

### PR DESCRIPTION
**Pull request by GANGANI**
### Purpose
With these PR changes, SSO to console through Myaccount issue will be resolved.

### Approach
With the previous implementation, it will check the already authenticated authenticator list and if the current step is already authenticated it will SSO to the application. But in the console, as we have configured, multi-option for the first step along with identifier-first, GitHub and Google and since we didn't authenticate from any of these from MyAccount it will prompt for multi-option page. Even though from My account, it will authenticate from the basic authenticator, the identifier first step will not include for the already authenticated category. Therefore identifier first will prompt the user along with the multi-option page.

With the PR changes, it will check whether the multi-option is available including an authentication flow handler. If so it will give priority to identifier first and execute, and if the authentication status is incomplete with identifier-first it will re-prompt for the multi-option page.

### Related issue
https://github.com/wso2-enterprise/asgardeo-product/issues/3717

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 1]
```diff
@@ -290,11 +290,22 @@ else if (ConfigurationFacade.getInstance().isDumbMode() && authenticatedIdPs.isE
...
                 // IdP. If yes, don't send to the multi-option page. Call directly.
                 boolean sendToPage = false;
+                boolean isAuthenticationFlowHandler = false;
                 AuthenticatorConfig authenticatorConfig = null;
 
                 // Are there multiple authenticators?
                 if (authConfigList.size() > 1) {
                     sendToPage = true;
+                    // To identify whether the multi-option is available including an authentication flow handler.
+                    // If it is available it will directly redirect to that.
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:300]: Shall we use multiline comments?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 2]
```diff
@@ -322,45 +333,59 @@ else if (ConfigurationFacade.getInstance().isDumbMode() && authenticatedIdPs.isE
...
+            throws FrameworkException {
+
+        String loginPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL();
+        if (log.isDebugEnabled()) {
+            log.debug("Sending to the Multi Option page");
+        }
+        diagnosticLog.info("Sending to the Multi Option page");
+        Map<String, String> parameterMap = getAuthenticatorConfig().getParameterMap();
+        String showAuthFailureReason = null;
+        if (parameterMap != null) {
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:364]: Shall we use `MapUtils`?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 3]
```diff
@@ -322,45 +333,59 @@ else if (ConfigurationFacade.getInstance().isDumbMode() && authenticatedIdPs.isE
...
+            showAuthFailureReason = parameterMap.get(FrameworkConstants.SHOW_AUTHFAILURE_RESON_CONFIG);
+            if (log.isDebugEnabled()) {
+                log.debug("showAuthFailureReason has been set as : " + showAuthFailureReason);
+            }
+        }
+        diagnosticLog.info("showAuthFailureReason has been set as : " + showAuthFailureReason);
+        String retryParam = "";
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:371]: `StringUtils.EMPTY` Wwdyt?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 2 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3533/commits/e2e1a48558bfa645abdcdeec06bba479bf62a72e

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 3 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3533/commits/e2e1a48558bfa645abdcdeec06bba479bf62a72e

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 4]
```diff
@@ -290,11 +290,22 @@ else if (ConfigurationFacade.getInstance().isDumbMode() && authenticatedIdPs.isE
                 // Find if step contains only a single authenticator with a single
                 // IdP. If yes, don't send to the multi-option page. Call directly.
                 boolean sendToPage = false;
+                boolean isAuthenticationFlowHandler = false;
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:293]:
```suggestion
                boolean isAuthenticationFlowHandlerInMultiStep = false;
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 5]
```diff
@@ -322,45 +333,59 @@ else if (ConfigurationFacade.getInstance().isDumbMode() && authenticatedIdPs.isE
                     }
 
                     doAuthentication(request, response, context, authenticatorConfig);
+                    /* If an authentication flow handler is redirected with incomplete, it will redirect to
+                    multi option page, as multi-option is available */
+                    if (context.getParameter(FrameworkConstants.REDIRECTED_AUTHENTICATOR) ==
+                            authenticatorConfig.getApplicationAuthenticator().getName()
+                            && isAuthenticationFlowHandler) {
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:340]:
```suggestion
                    if (request.getAttribute(FrameworkConstants.RequestParams.FLOW_STATUS) ==
                            AuthenticatorFlowStatus.INCOMPLETE
                            && isAuthenticationFlowHandlerInMultiStep) {
```

Can't we use this? with this, we don't need to add another property.

Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3533/commits/0005f836e531618f6c8b0c701cf6733750181ae3

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 4 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3533/commits/0005f836e531618f6c8b0c701cf6733750181ae3

**Review by senthalan — APPROVED** [review]

**Pull request merged**